### PR TITLE
Search Endpoint

### PR DESCRIPTION
**Pull request by HUmutOzdemir**
Please Make a code review and merge if there is not any problem. If there are sone things that must change please write on the linked issue about Search Endpoint.

Please control these criteria.(If you want you can add extra criteria.)
- [x] Documentation
- [x] Readability
- [x] Code works correctly or not?
- [x] Tests

### CODE REVIEW

**Comment by oykuyilmaz1** [comment]
I analyzed your code and maybe you can think of changing those:
- Your test.py does not have enough documentation I think, you have comments that clarifying the functions of your unit test but maybe you can add some definitions to that.
- in your searchEngine.py, you have a nice documentation and when I read I can see what the functions are doing clearly and it works. But you can review the error messages as follows: 
* Line 88 you have a typo --> "parameters"
* Line 100 --> "There is a problem about the search API. Please try again!!" (So that the user have more information of what went wrong)
* Line 241 --> "You have entered a wrong keyword, or selected wrong filters. Or maybe you have not logged in." (I think getparameters function also checks if the user has logged in if not please ignore it. Also you can improve the sentence ot correct it. The point is to be more specific to the user about the error.)
* Line 244 --> Maybe you have not logged in? (Same argument goes for this again.)
* Line 252 -->"There is a problem about the search API. Please try again!!" 
Other than that I think your code is readable, works correctly and well indented. Thank you for your effort.

**Comment by HUmutOzdemir** [comment]
Your feedback seems true. I will update the code until tomorrow. Is this okey for you?

**Comment by oykuyilmaz1** [comment]
Sure, it is fine. Sorry for the late response.

File: practice-app/platon_api/rest_api/search_engine/search_engine.py [hunk 1]
```diff
@@ -0,0 +1,295 @@
...
+    
+    @staticmethod
+    def sort_output(output_list,sorting_type):
+        """
+            where 'output_list': the list of the results that come without sorting
+            where 'sorting_type': sorting type parameter
+
+            returns sorted version of the search result
+
+            This function sorts the output with a given sotring criteria
```
Comment (ertbulbul) [inline practice-app/platon_api/rest_api/search_engine/search_engine.py:169]: sorting misspelled

File: practice-app/platon_api/rest_api/search_engine/documentation/search_engine_documentation.md [hunk 2]
```diff
@@ -0,0 +1,126 @@
...
+     |      list of weak references to the object (if defined)
+     |
+     |  ----------------------------------------------------------------------
+     |  Data and other attributes defined here:
+     |
+     |  exact_match_score = 1000000
+     |
+     |  num_of_semantically_related = 5
+     |
+     |  sotring_criteria_list = ['name_increasing', 'name_decreasing', 'surnam...
```
Comment (ertbulbul) [inline practice-app/platon_api/rest_api/search_engine/documentation/search_engine_documentation.md:116]: sorting misspelled

File: practice-app/platon_api/rest_api/search_engine/documentation/search_engine_documentation.md [hunk 3]
```diff
@@ -0,0 +1,126 @@
...
+     |      returns a list of tuples which contails (word,semantic relation point)
+     |
+     |
+     |  sort_output(output_list, sorting_type)
+     |      where 'output_list': the list of the results that come without sorting
+     |      where 'sorting_type': sorting type parameter
+     |
+     |      returns sorted version of the search result
+     |
+     |      This function sorts the output with a given sotring criteria
```
Comment (ertbulbul) [inline practice-app/platon_api/rest_api/search_engine/documentation/search_engine_documentation.md:91]: sorting misspelled

**Comment by HUmutOzdemir** [comment]
> I analyzed your code and maybe you can think of changing those:
> 
> * Your test.py does not have enough documentation I think, you have comments that clarifying the functions of your unit test but maybe you can add some definitions to that.
> * in your searchEngine.py, you have a nice documentation and when I read I can see what the functions are doing clearly and it works. But you can review the error messages as follows:
> 
> * Line 88 you have a typo --> "parameters"
> * Line 100 --> "There is a problem about the search API. Please try again!!" (So that the user have more information of what went wrong)
> * Line 241 --> "You have entered a wrong keyword, or selected wrong filters. Or maybe you have not logged in." (I think getparameters function also checks if the user has logged in if not please ignore it. Also you can improve the sentence ot correct it. The point is to be more specific to the user about the error.)
> * Line 244 --> Maybe you have not logged in? (Same argument goes for this again.)
> * Line 252 -->"There is a problem about the search API. Please try again!!"
>   Other than that I think your code is readable, works correctly and well indented. Thank you for your effort.

I added extra documentation to the unit tests. Also I changed the error messages that will be seen but the user. Other strings that I returned in the get_parameters functions are some messages for me actually. These messages will not be seen by the user.

File: practice-app/platon_api/rest_api/search_engine/search_engine.py [hunk 4]
```diff
@@ -0,0 +1,295 @@
...
+        except:
+            return [(token,searchEngine.exact_match_score) for token in search_tokens]
+
+
+    @staticmethod
+    def get_parameter_list(request=None):
+        """
+            where 'request': HTTP request that comes from the view class
+
+            returns a dictionary of the paramters if there is no problem about the paramters
```
[the remark this answers is not in the document]
Comment (HUmutOzdemir) [reply practice-app/platon_api/rest_api/search_engine/search_engine.py]: I changed this

File: practice-app/platon_api/rest_api/search_engine/search_engine.py [hunk 5]
```diff
@@ -0,0 +1,295 @@
...
+            # Take the search string from URL
+            search_string = request.GET["search_string"].lower()
+            # Make search string only a empty string to return all users
+            if search_string == "":
+                search_string = " "
+        else:
+            return "Wrong input. Please give an appropriate input!!"
+        if request.GET.get("sorting_criteria",False):
+            # Take sorting criteria from URL
+            sotring_criteria = request.GET.get("sorting_criteria",False).replace('"',"")
```
[the remark this answers is not in the document]
Comment (HUmutOzdemir) [reply practice-app/platon_api/rest_api/search_engine/search_engine.py:129]: I changed this also

File: practice-app/platon_api/rest_api/search_engine/search_engine.py [hunk 6]
```diff
@@ -0,0 +1,295 @@
...
+            if sotring_criteria not in searchEngine.sotring_criteria_list:
```
[the remark this answers is not in the document]
Comment (HUmutOzdemir) [reply practice-app/platon_api/rest_api/search_engine/search_engine.py:130]: Done

File: practice-app/platon_api/rest_api/search_engine/search_engine.py [hunk 7]
```diff
@@ -0,0 +1,295 @@
...
+                return "Wrong input. Please give an appropriate input!!"
+        else:
+            # Default sorting criteria
+            sotring_criteria = ""
```
[the remark this answers is not in the document]
Comment (HUmutOzdemir) [reply practice-app/platon_api/rest_api/search_engine/search_engine.py:134]: Done

File: practice-app/platon_api/rest_api/search_engine/search_engine.py (hunk 1 above)
Comment (HUmutOzdemir) [reply practice-app/platon_api/rest_api/search_engine/search_engine.py:169]: Done

File: practice-app/platon_api/rest_api/search_engine/documentation/search_engine_documentation.md (hunk 2 above)
Comment (HUmutOzdemir) [reply practice-app/platon_api/rest_api/search_engine/documentation/search_engine_documentation.md:116]: Done

File: practice-app/platon_api/rest_api/search_engine/documentation/search_engine_documentation.md (hunk 3 above)
Comment (HUmutOzdemir) [reply practice-app/platon_api/rest_api/search_engine/documentation/search_engine_documentation.md:91]: Done

**Comment by HUmutOzdemir** [comment]
> I couldn't find an error other than simple spelling mistakes. Everything looks good. Well done

Thank you for your feedback. I changed all of my misspelled words.

**Pull request merged**